eva-usage-stats/ftp_usage.py

```diff
@@ -0,0 +1,145 @@
+    with get_metadata_connection_handle('development', private_config_xml_file) as metadata_connection_handle:
+        query_create_table = (
+            'CREATE TABLE IF NOT EXISTS eva_web_srvc_stats.ftp_traffic '
+            '(event_ts_txt TEXT, event_ts TIMESTAMP, host TEXT, uhost TEXT,'
+            ' request_time TEXT, request_year INTEGER, request_ts TIMESTAMP'
+            ' file_name TEXT, file_size INTEGER, transfer_time INTEGER,'
+            ' transfer_type CHAR, direction CHAR, special_action CHAR(4), access_mode CHAR,'
+            ' country CHAR(2), region TEXT, city TEXT, domain_name TEXT, isp TEXT, usage_type TEXT)'
+        )
+    execute_query(metadata_connection_handle, query_create_table)
+
+
+def load_batch_to_table(batch, private_config_xml_file):
+    logger.info(f'Loading {len(batch)} records...')
+    rows = [(
+        b['@timestamp'],  # event timestamp
+        b['@timestamp'],  # to be converted
+        b['host'],  # webprod host
+        b['uhost'],  # unique user host string
+        # FTP log fields: see https://docs.oracle.com/cd/E19683-01/817-0667/6mgevq0ee/index.html
+        b['current_time'],
+        b['year'],
+        f"{b['year']} {b['current_time']}",  # to be converted
+        b['file_name'],
+        b['file_size'],
+        b['transfer_time'],
+        b['transfer_type'],
+        b['direction'],
+        b['special_action'],
+        b['access_mode'],
+        # IP2Location fields: see https://www.ip2location.com/web-service/ip2location
+        b['ip2location']['country_short'],
+        b['ip2location']['region'],
+        b['ip2location']['city'],
+        b['ip2location']['domain'],
+        b['ip2location']['isp'],
+        b['ip2location']['usage_type'],
+    ) for b in batch]
+    with get_metadata_connection_handle('development', private_config_xml_file) as metadata_connection_handle:
+        with metadata_connection_handle.cursor() as cursor:
+            query_insert = (
+                'INSERT INTO eva_web_srvc_stats.ftp_traffic '
+                'VALUES (%s, cast(%s as timestamp with time zone), %s, %s, %s, %s, '
+                'cast(%s as timestamp without time zone), %s, %s, %s, %s, %s, %s, '
+                '%s, %s, %s, %s, %s, %s, %s)'
+            )
+            psycopg2.extras.execute_batch(cursor, query_insert, rows)
+
+
+def get_most_recent_timestamp(private_config_xml_file):
+    with get_metadata_connection_handle('development', private_config_xml_file) as metadata_connection_handle:
+        results = get_all_results_for_query(
+            metadata_connection_handle,
+            "select max(event_ts_txt) as recent_ts from eva_web_srvc_stats.ftp_traffic;"
+        )
+        if results and results[0][0]:
+            return results[0][0]
+    return None
+
+
+@retry(tries=4, delay=2, backoff=1.2, jitter=(1, 3))
+def query(kibana_host, basic_auth, private_config_xml_file):
+    first_query_url = os.path.join(kibana_host, 'ftplogs*/_search?scroll=24h')
```

Comment:
I think this query will be fine but we might get a faster query if we limit the search space to a single (or a few) index in line with the `most_recent_timestamp`. We can revisit if we need to.